### ISSUE TITLE
Research and document incentivisation in the KSN network

### ISSUE THREAD

**Issue by dryajov** [MetaMask/kitsunet-docs#4]
# Incentivization in KSN

## TL;DR

Incentivization serves two purposes 

- [ ] Data availability in the network
- [ ] Additional security layer for detering bad behavior

### Requirements

- Incentivize Ethereum full node participation
- Incentivize data availability in KSN network
- Incentivize sharing behaviour in the KSN network
  - nodes should be compensated for `seeding` to other nodes
  - nodes should be `charged` for `leeching` from other nodes

#### Incentivize Ethereum full node participation

The KSN network is a p2p network of light ethereum clients that collaborate in distributing slices (merkle subtries) of the ethereum state. Currently, the KSN network doesn't have any sort of incentivization layer that would ensure the availability of the ethereum state. Furthermore, there are currently no direct incentives in the Ethereum network itself to run full nodes and ensure data availability to other full and light nodes. This is a problem, because unless an actor is directly invested in the Ethereum network (through mining, products and services, or by holding significant amounts of ETH), there is no motivation for a casual user to run a full node, moreover, there is no motivation for infrastructure providers to get involved in the Ethereum network unless they are again directly invested in it.

**One of the goals of this issue is to look for ways of creating an incentivization network to reward and encourage full node participation in the Ethereum network.**

#### Incentivize data availability in KSN network 

The KSN network depends on full Ethereum nodes allow timely access to up to date and historical Ethereum state. KSN nodes create an overlay mesh where pieces of the ethereum state get propagated, this state comes from full Ethereum nodes that are willing to serve it to KSN (or any other) nodes. Without any incentives for providing this data, full Ethereum nodes will quickly block KSN clients for `leeching` onto them, hence, a way of compensating full nodes serving other nodes with data is required.

This requirement might be covered by the previous one (`Incentivize Ethereum full node participation`), but its left as a separate entry to encourage further discussion.

#### Incentivize sharing behaviour in the KSN network

KSN nodes `seed` data to other KSN nodes - this is a fundamental requirement of the Mustekala project. We need to develop a way where nodes that `seed` data to other nodes are compensated and those that `leech` are penalized, this is to ensure equal participation of _ALL_ the nodes in the network. 

The most basic approach is a `tit-for-tat` system, however, it falls short in many cases. Furthermore, we believe that there are many missed opportunities with a simplistic approach such as `tit-for-tat` that ignores all the possibilities that tokenized incentives bring to the table.

We would like to design a protocol where:

- `seeding` peers are compensated for the data and bandwidth they provide
- `leeching` peers are penalized/charged for the data and bandwidth they consume
